File: 백준/Silver/5567. 결혼식/결혼식.js
const input = require("fs")
  .readFileSync("/dev/stdin", "utf8")
  .trim()
  .split("\n");

// 첫번째줄에 사람수, 두번째줄에 관계수다
// 3번째부터 관계가 주어지고 양방향이다
// 1로 연결된 모든 사람의 수를 구하면된다
// bfs로 탐색하면서 visited로 마킹하면된다
const solution = (input) => {
  const peopleN = Number(input[0]);
  const lineN = Number(input[1]);
  const graph = Array(peopleN + 1)
    .fill(0)
    .map((v) => []);

  for (let i = 2; i < 2 + lineN; i++) {
    const [n1, n2] = input[i].split(" ").map((v) => Number(v));
    graph[n1].push(n2);
    graph[n2].push(n1);
  }

  const visited = Array(peopleN + 1).fill(false);
  visited[1] = true;

  const queue = [[1, 0]];

  while (queue.length) {
    const [curPerson, relation] = queue.shift();
    if (relation > 1) break;

    for (const nextPerson of graph[curPerson]) {
      if (visited[nextPerson]) continue;
      queue.push([nextPerson, relation + 1]);
      visited[nextPerson] = true;
    }
  }

  console.log(visited.filter((v) => v === true).length - 1);
};

solution(input);

File: 백준/Silver/5567. 결혼식/결혼식.py
# 1부터 N명의 사람들중에서 친구와 친구의 친구까지 초대한다
# 첫째줄에 동기수 둘째줄에 리스트이 길이 그다음줄 부터 관계가 주어진다
# 초대할 동기수를 구해라
# bfs로 [1, 0]부터 시작해서 관계값이 2보다 크면 break하면된다
from collections import deque

peopleN = int(input())
relationN = int(input())
grid = [[] for _ in range(peopleN+1)]

for i in range(relationN):
  n1, n2 = list(map(int, input().split()))
  grid[n1].append(n2)
  grid[n2].append(n1)

queue = deque()
queue.append([1, 0])
peopleDist = [float('inf')]*(peopleN+1)
peopleDist[1] = 0
answer = 0 

while len(queue):
  person, dist = queue.popleft()
  if dist > 2:
    break

  if 1<=dist<=2:
    answer += 1

  for nextPerson in grid[person]:
    if peopleDist[nextPerson] > dist+1:
      peopleDist[nextPerson] = dist + 1
      queue.append([nextPerson, dist + 1])

print(answer)